### PR TITLE
HTTP2: Always eagerly fire frames to the child channels but manage ba…

### PR DESCRIPTION
**Pull request by normanmaurer**
…ckpressure via window update frames by read()

Motivation:

At the moment each Http2StreamChannel instance has a complicated state machine to handle auto read configurations. We can simplify this by just always fire frames to the child channel pipelines as soon as we receive them but only give back bytes to the flowcontroller when read() is called. This will ensure that backpressure still is correctly managed via flow control (by window update frames) but simplify things a lot and may release memory more quickly.

Fore more details see https://github.com/netty/netty/pull/9389#issuecomment-512891869.

Special thanks to @bryce-anderson for the idea.

Modifications:

- Remove all the complex buffering logic from AbstractHttp2StreamChannel
- Move logic to give back bytes to the flow-controller (and so produce window update frames) to beginRead(...)
- Update tests

Result:

Less complex implementation while still maintain ability to flow control channels.

### CODE REVIEW

**Comment by normanmaurer** [comment]
This is based on the idea that was brought up by @bryce-anderson and I mainly opened it to start discussions around the idea. I actually like it a lot, but will do some more testing / benchmarking tomorrow to verify it in more depth.

**Comment by bryce-anderson** [comment]
I'll pull the patch and take look later this afternoon, but this looks even better than I'd imagined.

**Comment by ejona86** [comment]
If AUTO_READ is off, I'm trying to see how this plays out. Normally code would wait until onReadComplete to decide if it wanted to read more, and if not would call read() when it wanted more. Now onReadComplete will be called multiple times, even in the absence of read()s. That seems like asking for trouble. (And avoiding onReadComplete but still having channelRead also seems to be trouble.)

The previous code appears to be been broken in the same way, as it would deliver these extra onReadComplete()s, even if it didn't deliver any channelReads. But that could be fixed.

Even without the larger changes, it looks like we could still move updateLocalWindowIfNeeded() to doBeginRead to reduce the memory used.

I do agree this code is simpler, but it's not clear to me we can get away with it.

File: codec-http2/src/main/java/io/netty/handler/codec/http2/AbstractHttp2StreamChannel.java [hunk 1]
```diff
@@ -543,17 +509,11 @@ void fireChildRead(Http2Frame frame) {
             } else {
                 unsafe.notifyReadComplete(allocHandle, true);
```
Comment (bryce-anderson) [inline codec-http2/src/main/java/io/netty/handler/codec/http2/AbstractHttp2StreamChannel.java]: Now that we're essentially eager-reading, why don't we always trigger read complete via `addChannelToReadCompletePendingQueue();`? This should avoid about as many `readComplete` calls as we can.
As an aside, It seems to me that we shouldn't care much what `allocHandle.continueReading()` says at this point, or honestly at all with this strategy.

**Comment by normanmaurer** [comment]
@bryce-anderson auto read just calls `read()` automatically so `beginRead()` is called as well. Which means it does not matter if its auto read or not it should work the same way. I am missing anything ?

**Comment by normanmaurer** [comment]
@ejona86 I don't share your concerns honestly. `read()` is just a hint that we want to read more data, how much is dispatched is a different matter. There is no 1:1 mapping between `read() `and `channelRead(...)`, as even one `read()` can produce multiple `channelRead(...)` as of today (like when the `ByteToMesssageDecoder` is used. What I am missing ?

File: codec-http2/src/main/java/io/netty/handler/codec/http2/AbstractHttp2StreamChannel.java (hunk 1 above)
Comment (normanmaurer) [reply codec-http2/src/main/java/io/netty/handler/codec/http2/AbstractHttp2StreamChannel.java]: @bryce-anderson good point... let me do this.

**Comment by normanmaurer** [comment]
FYI I did some benchmarks after this PR was applied and the speed is the same.

**Comment by ejona86** [comment]
> There is no 1:1 mapping between read() and channelRead(...), as even one read() can produce multiple channelRead(...) as of today

I'm not bothered by that. I'm bothered that the channel will see _channelReadComplete_ multiple times.

**Comment by normanmaurer** [comment]
@ejona86 with the latest adjustments it will only see it once... Can you check again ?

**Comment by normanmaurer** [comment]
@netty-bot test this please

**Comment by ejona86** [comment]
> @ejona86 with the latest adjustments it will only see it once... Can you check again ?

Looks like it still happens multiple times. Yes, there will be channelReads between each channelReadComplete, but channelReadComplete will not be 1:1 with read().

Look at it this way: channelReadComplete() can be called even if the read() was never called.

**Comment by normanmaurer** [comment]
@ejona86 that is also true for many other handlers... There is no mapping between `read()` and `channelReadComplete` imho.... the important thing is that if there is a `channelRead` there also must be a `channelReadComplete`.

**Comment by normanmaurer** [comment]
@ejona86 @bryce-anderson @trustin WDYT ?

**Comment by bryce-anderson** [comment]
So I really like this because it's so much simpler and avoids duplicating the buffering that the members of the channel are going to have to inevitably do anyway.

If I understand this all correctly, with regard to unexpected `channelReadComplete` calls, before 60cf18cf2027 an EOS frame would trigger a flush of the buffer so we'd potentially get frames and a call to `channelReadComplete` without a `.read()` request. I believe that means it was necessary to be able to field all messages regardless of whether `.read()` was call. That said, the documentation on `channelReadComplete` could make this surprising behavior:
```java
/**
     * ...  If {@link ChannelOption#AUTO_READ} is off, no further
     * attempt to read an inbound data from the current {@link Channel} will be made until
     * {@link ChannelHandlerContext#read()} is called.
     */
    void channelReadComplete(ChannelHandlerContext ctx) throws Exception;
```

So I'm a weak 👍 on this because feels clean, efficient, and this behavior is something people had to deal with before but I also hear @ejona86's point that this is, at best, a corner case in the event model contract that _could_ surprise people.

**Comment by normanmaurer** [comment]
let us close this one for now.

**Pull request closed without merging**